// @xmtp/xmtp-js/dist/cjs/src/crypto/Signature.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.StaticWalletAccountLinkSigner = exports.SIWEWalletAccountLinkSigner = exports.WalletSigner = exports.AccountLinkedRole = exports.AccountLinkedSIWESignature = exports.AccountLinkedStaticSignature = exports.ecdsaSignerKey = void 0;
const proto_1 = require("@xmtp/proto");
const long_1 = __importDefault(require("long"));
const secp = __importStar(require("@noble/secp256k1"));
const PublicKey_1 = require("./PublicKey");
const PrivateKey_1 = require("./PrivateKey");
const ethers_1 = require("ethers");
const utils_1 = require("./utils");
const utils_2 = require("ethers/lib/utils");
const siwe_1 = require("siwe");
// Validate signature.
function ecdsaCheck(sig) {
    if (sig.bytes.length !== 64) {
        throw new Error(`invalid signature length: ${sig.bytes.length}`);
    }
    if (sig.recovery !== 0 && sig.recovery !== 1) {
        throw new Error(`invalid recovery bit: ${sig.recovery}`);
    }
}
// Compare signatures.
function ecdsaEqual(a, b) {
    return a.recovery === b.recovery && (0, utils_1.equalBytes)(a.bytes, b.bytes);
}
// Derive public key of the signer from the digest and the signature.
function ecdsaSignerKey(digest, signature) {
    const bytes = secp.recoverPublicKey(digest, signature.bytes, signature.recovery);
    return bytes
        ? new PublicKey_1.UnsignedPublicKey({
            secp256k1Uncompressed: { bytes },
            createdNs: long_1.default.fromNumber(0),
        })
        : undefined;
}
exports.ecdsaSignerKey = ecdsaSignerKey;
class Signature {
    constructor(obj) {
        if (obj.ecdsaCompact) {
            ecdsaCheck(obj.ecdsaCompact);
            this.ecdsaCompact = obj.ecdsaCompact;
        }
        else if (obj.walletEcdsaCompact) {
            ecdsaCheck(obj.walletEcdsaCompact);
            this.walletEcdsaCompact = obj.walletEcdsaCompact;
        }
        else {
            throw new Error('invalid signature');
        }
    }
    // Return the public key that validates provided key's signature.
    signerKey(key) {
        return __awaiter(this, void 0, void 0, function* () {
            if (this.ecdsaCompact) {
                return PrivateKey_1.SignedPrivateKey.signerKey(key, this.ecdsaCompact);
            }
            else if (this.walletEcdsaCompact) {
                return WalletSigner.signerKey(key, this.walletEcdsaCompact);
            }
            else {
                return undefined;
            }
        });
    }
    // LEGACY: Return the public key that validates this signature given the provided digest.
    // Return undefined if the signature is malformed.
    getPublicKey(digest) {
        if (!this.ecdsaCompact) {
            throw new Error('invalid signature');
        }
        const bytes = secp.recoverPublicKey(digest, this.ecdsaCompact.bytes, this.ecdsaCompact.recovery);
        return bytes
            ? new PublicKey_1.PublicKey({
                secp256k1Uncompressed: { bytes },
                timestamp: long_1.default.fromNumber(0),
            })
            : undefined;
    }
    // Is this the same/equivalent signature as other?
    equals(other) {
        if (this.ecdsaCompact && other.ecdsaCompact) {
            return ecdsaEqual(this.ecdsaCompact, other.ecdsaCompact);
        }
        if (this.walletEcdsaCompact && other.walletEcdsaCompact) {
            return ecdsaEqual(this.walletEcdsaCompact, other.walletEcdsaCompact);
        }
        return false;
    }
    toBytes() {
        return proto_1.signature.Signature.encode(this).finish();
    }
    static fromBytes(bytes) {
        return new Signature(proto_1.signature.Signature.decode(bytes));
    }
}
exports.default = Signature;
class AccountLinkedStaticSignatureV1 {
    constructor(obj) {
        if (!obj.text || !obj.signature) {
            throw new Error('Invalid AccountLinkedStaticSignatureV1');
        }
        this.text = obj.text;
        this.signature = new Signature(obj.signature);
        if (!this.signature.walletEcdsaCompact) {
            throw new Error('Invalid AccountLinkedStaticSignatureV1 does not have ecdsaCompact');
        }
        this._walletEcdsaCompact = this.signature.walletEcdsaCompact;
    }
    get walletEcdsaCompact() {
        return this._walletEcdsaCompact;
    }
}
class AccountLinkedStaticSignature extends AccountLinkedStaticSignatureV1 {
    constructor(obj) {
        if (!obj.v1) {
            throw new Error('Unsupported AccountLinkedStaticSignature version');
        }
        super(obj.v1);
        this.v1 = obj.v1;
    }
    static create(text, signature) {
        return new AccountLinkedStaticSignature({
            v1: {
                text,
                signature,
            },
        });
    }
}
exports.AccountLinkedStaticSignature = AccountLinkedStaticSignature;
class AccountLinkedSIWESignatureV1 {
    constructor(obj) {
        if (!obj.text || !obj.signature) {
            throw new Error('Invalid AccountLinkedSIWESignatureV1');
        }
        this.text = obj.text;
        this.signature = new Signature(obj.signature);
        if (!this.signature.walletEcdsaCompact) {
            throw new Error('Invalid AccountLinkedSIWESignatureV1 does not have ecdsaCompact');
        }
        this._walletEcdsaCompact = this.signature.walletEcdsaCompact;
    }
    get walletEcdsaCompact() {
        return this._walletEcdsaCompact;
    }
}
class AccountLinkedSIWESignature extends AccountLinkedSIWESignatureV1 {
    constructor(obj) {
        if (!obj.v1) {
            throw new Error('Unsupported AccountLinkedSIWESignature version');
        }
        super(obj.v1);
        this.v1 = obj.v1;
    }
    static create(text, signature) {
        return new AccountLinkedSIWESignature({
            v1: {
                text,
                signature,
            },
        });
    }
}
exports.AccountLinkedSIWESignature = AccountLinkedSIWESignature;
var AccountLinkedRole;
(function (AccountLinkedRole) {
    AccountLinkedRole[AccountLinkedRole["INBOX_KEY"] = 0] = "INBOX_KEY";
    AccountLinkedRole[AccountLinkedRole["SEND_KEY"] = 1] = "SEND_KEY";
})(AccountLinkedRole = exports.AccountLinkedRole || (exports.AccountLinkedRole = {}));
// A wallet based KeySigner.
class WalletSigner {
    constructor(wallet) {
        this.wallet = wallet;
    }
    static identitySigRequestText(keyBytes) {
        // Note that an update to this signature request text will require
        // addition of backward compatibility for existing signatures
        // and/or a migration; otherwise clients will fail to verify previously
        // signed keys.
        return ('XMTP : Create Identity\n' +
            `${(0, utils_1.bytesToHex)(keyBytes)}\n` +
            '\n' +
            'For more info: https://xmtp.org/signatures/');
    }
    static signerKey(key, signature) {
        const digest = (0, utils_1.hexToBytes)(ethers_1.utils.hashMessage(this.identitySigRequestText(key.bytesToSign())));
        return ecdsaSignerKey(digest, signature);
    }
    signKey(key) {
        return __awaiter(this, void 0, void 0, function* () {
            const keyBytes = key.toBytes();
            const sigString = yield this.wallet.signMessage(WalletSigner.identitySigRequestText(keyBytes));
            const eSig = ethers_1.utils.splitSignature(sigString);
            const r = (0, utils_1.hexToBytes)(eSig.r);
            const s = (0, utils_1.hexToBytes)(eSig.s);
            const sigBytes = new Uint8Array(64);
            sigBytes.set(r);
            sigBytes.set(s, r.length);
            const signature = new Signature({
                walletEcdsaCompact: {
                    bytes: sigBytes,
                    recovery: eSig.recoveryParam,
                },
            });
            return new PublicKey_1.SignedPublicKey({ keyBytes, signature });
        });
    }
}
exports.WalletSigner = WalletSigner;
class SIWEWalletAccountLinkSigner {
    constructor(wallet) {
        this.wallet = wallet;
    }
    static accountLinkedSIWERoleRequestText(role) {
        switch (role) {
            case AccountLinkedRole.INBOX_KEY:
                return 'AllowAllRead';
            case AccountLinkedRole.SEND_KEY:
                return 'GrantSendPermissions';
        }
    }
    static accountLinkedSIWEResourceRoleText(keyBytes, role) {
        return `https://xmtp.org/siwe/${SIWEWalletAccountLinkSigner.accountLinkedSIWERoleRequestText(role)}/secp256k1/${(0, utils_1.bytesToBase64)(keyBytes)}`;
    }
    // Default SIWE text to be signed, most apps will NOT want to use this.
    // Most apps will want to include the resource string in their own SIWE to avoid
    // asking for multiple SIWE signatures.
    static defaultAccountLinkSIWERequestText(key, role, walletAddress) {
        // Create a SIWE message
        // - statement can be anything
        // - get the address from the signer
        // - add the resource string with keybytes
        const keyBytes = key.toBytes();
        const resource = SIWEWalletAccountLinkSigner.accountLinkedSIWEResourceRoleText(keyBytes, role);
        const siwe = new siwe_1.SiweMessage({
            statement: 'XMTP Account Link with Role: ' + role,
            address: walletAddress,
            domain: 'xmtp.org',
            version: '1',
            uri: 'https://xmtp.org',
            chainId: 1,
            resources: [resource],
        });
        return siwe.prepareMessage();
    }
    // NOTE: this will NOT be used in practice, mostly here for demonstration purposes
    // in reality, the SDK consumer will likely use their login SIWE with the role resource
    // and pass it directly down into the client
    signKeyWithRole(key, role) {
        return __awaiter(this, void 0, void 0, function* () {
            const keyBytes = key.toBytes();
            const siweMessageBytes = (0, utils_2.toUtf8Bytes)(SIWEWalletAccountLinkSigner.defaultAccountLinkSIWERequestText(key, role, yield this.wallet.getAddress()));
            const sigString = yield this.wallet.signMessage(siweMessageBytes);
            const eSig = ethers_1.utils.splitSignature(sigString);
            const r = (0, utils_1.hexToBytes)(eSig.r);
            const s = (0, utils_1.hexToBytes)(eSig.s);
            const sigBytes = new Uint8Array(64);
            sigBytes.set(r);
            sigBytes.set(s, r.length);
            const signature = new Signature({
                walletEcdsaCompact: {
                    bytes: sigBytes,
                    recovery: eSig.recoveryParam,
                },
            });
            return PublicKey_1.AccountLinkedPublicKey.create(keyBytes, undefined, AccountLinkedSIWESignature.create(siweMessageBytes, signature));
        });
    }
}
exports.SIWEWalletAccountLinkSigner = SIWEWalletAccountLinkSigner;
class StaticWalletAccountLinkSigner {
    constructor(wallet) {
        this.wallet = wallet;
    }
    static accountLinkedRoleRequestText(role) {
        switch (role) {
            case AccountLinkedRole.INBOX_KEY:
                return 'Create Identity';
            case AccountLinkedRole.SEND_KEY:
                return 'Grant Send Permissions';
        }
    }
    static accountLinkRequestText(keyBytes, role) {
        // Note that an update to this signature request text will require
        // addition of backward compatibility for existing signatures
        // and/or a migration; otherwise clients will fail to verify previously
        // signed keys.
        return (`XMTP : ${StaticWalletAccountLinkSigner.accountLinkedRoleRequestText(role)}\n` +
            `${(0, utils_1.bytesToHex)(keyBytes)}\n` +
            '\n' +
            'For more info: https://xmtp.org/signatures/');
    }
    signKeyWithRole(key, role) {
        return __awaiter(this, void 0, void 0, function* () {
            const keyBytes = key.toBytes();
            const requestTextBytes = (0, utils_2.toUtf8Bytes)(StaticWalletAccountLinkSigner.accountLinkRequestText(keyBytes, role));
            const sigString = yield this.wallet.signMessage(requestTextBytes);
            const eSig = ethers_1.utils.splitSignature(sigString);
            const r = (0, utils_1.hexToBytes)(eSig.r);
            const s = (0, utils_1.hexToBytes)(eSig.s);
            const sigBytes = new Uint8Array(64);
            sigBytes.set(r);
            sigBytes.set(s, r.length);
            const signature = new Signature({
                walletEcdsaCompact: {
                    bytes: sigBytes,
                    recovery: eSig.recoveryParam,
                },
            });
            return PublicKey_1.AccountLinkedPublicKey.create(keyBytes, AccountLinkedStaticSignature.create(requestTextBytes, signature), undefined /* siweSignature */);
        });
    }
}
exports.StaticWalletAccountLinkSigner = StaticWalletAccountLinkSigner;
