// @xmtp/xmtp-js/dist/cjs/src/utils/topic.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.buildUserPrivateStoreTopic = exports.buildUserInviteTopic = exports.buildUserIntroTopic = exports.buildUserContactTopic = exports.buildDirectMessageTopicV2 = exports.buildDirectMessageTopic = exports.buildContentTopic = void 0;
const ethers_1 = require("ethers");
const buildContentTopic = (name) => `/xmtp/0/${name}/proto`;
exports.buildContentTopic = buildContentTopic;
const buildDirectMessageTopic = (sender, recipient) => {
    // EIP55 normalize the address case.
    const members = [ethers_1.utils.getAddress(sender), ethers_1.utils.getAddress(recipient)];
    members.sort();
    return (0, exports.buildContentTopic)(`dm-${members.join('-')}`);
};
exports.buildDirectMessageTopic = buildDirectMessageTopic;
const buildDirectMessageTopicV2 = (randomString) => {
    return (0, exports.buildContentTopic)(`m-${randomString}`);
};
exports.buildDirectMessageTopicV2 = buildDirectMessageTopicV2;
const buildUserContactTopic = (walletAddr) => {
    // EIP55 normalize the address case.
    return (0, exports.buildContentTopic)(`contact-${ethers_1.utils.getAddress(walletAddr)}`);
};
exports.buildUserContactTopic = buildUserContactTopic;
const buildUserIntroTopic = (walletAddr) => {
    // EIP55 normalize the address case.
    return (0, exports.buildContentTopic)(`intro-${ethers_1.utils.getAddress(walletAddr)}`);
};
exports.buildUserIntroTopic = buildUserIntroTopic;
const buildUserInviteTopic = (walletAddr) => {
    // EIP55 normalize the address case.
    return (0, exports.buildContentTopic)(`invite-${ethers_1.utils.getAddress(walletAddr)}`);
};
exports.buildUserInviteTopic = buildUserInviteTopic;
const buildUserPrivateStoreTopic = (addrPrefixedKey) => {
    // e.g. "0x1111111111222222222233333333334444444444/key_bundle"
    return (0, exports.buildContentTopic)(`privatestore-${addrPrefixedKey}`);
};
exports.buildUserPrivateStoreTopic = buildUserPrivateStoreTopic;
